Train the trainer (#236)
ome/omero-guides: training.rst

Prepare a training on OMERO
===========================

This chapter describes a setup of a training OMERO.server at 
your institution. Further, it gives an overview of the other training
resources as well as hints for trainers about how to present OMERO during a training session.

Resources
---------

- Community contributions and setups: `Github issue <https://github.com/ome/omero-guides/issues/107>`_.

- `Useful scripts <https://github.com/ome/training-scripts>`_ for setting up the training server.

Setup of scripts and OMERO.cli environment
------------------------------------------

- This guide assumes that you have installed an OMERO.server for training purposes.

- Clone the `training-scripts <https://github.com/ome/training-scripts>`_ repository::

>     $ git clone https://github.com/ome/training-scripts.git

Reviewer: The heading "Setup of scripts" made me expect we'd install them on the server or something, but I see you're just downloading them for use below. So maybe "Download of scripts and OMERO.cli environment setup" ?

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/68522283f4402ea8334b450b705ed8a77c180fc1

- Set up a OMERO.cli as specified under `CLI installation <https://docs.openmicroscopy.org/omero/latest/users/cli/installation.html>`_. Typically, this environment will be used on your local machine. Alternatively, you can use the OMERO.cli environment of the OMERO.server.

Setup of groups and users
-------------------------

Prepare groups and users as listed in the `provided template sheet <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/create_groups_users_setup>`_. Assuming you have the `template sheet <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/create_groups_users_setup>`_ in the same directory as the script `create_groups_users.sh <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/create_groups_users.sh>`_, you can run::

>     $ HOST=$YOUR_SERVER_ADDRESS PASSWORD=$PASSWORD_FOR_ROOT bash create_groups_users.sh

Reviewer: Could add a `$ cd training-scripts/maintenance/scripts` line before this, especially since users reading the docs won't see the full URLs of the scripts you mention.

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/68522283f4402ea8334b450b705ed8a77c180fc1

which will create 50 users in your database.
The users are members of four main OMERO.groups, which cover
the group permissions allowed in OMERO. The ``Read-annotate`` group ``Lab 1`` is the main group used in the trainings,
which contains most of the data in a typical training. The setup in the `template sheet <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/create_groups_users_setup>`_ sets this ``Lab 1`` group as a default group in OMERO for all users.

> Rename users to have first and last names of real people (the list of famous scientist names is used) running `the renaming script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/rename_users.py>`_.

Reviewer: Maybe show how this is run (as you do with other scripts). 
E.g. 
```
# login credentials will be requested
$ python rename_users.py
```

Import images
-------------

> `Import data in-place <https://omero-guides.readthedocs.io/en/latest/upload/docs/import-cli.html#in-place-import-using-the-cli>`_ for all users using `import bash script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/in_place_import_as.sh>`_ and the `publicly available data provided <https://downloads.openmicroscopy.org/images/>`_ or use your own data or `data downloaded from IDR <https://idr.openmicroscopy.org/about/download.html>`_. Shell into the machine where you have installed your OMERO.server (necessary for in-place import) and, assuming that you have for example your data in a folder `my-folder` which is visible from that machine::

Reviewer: Possibly this section could use an intro paragraph: "We like to use [in-place import] because it avoids duplicating the data on disk, but if you are only importing a small amount of data you may find non-in-place import (described below) is more convenient"?  (I guess "convenient" depends on their setup, but it seems trickier to download the data into a suitable place on the server?)

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/68522283f4402ea8334b450b705ed8a77c180fc1

> Prepare a training on OMERO

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/ea7d1e24beaa4617ab1ce5e540624064afb32b3d

    $ python rename_users.py trainer-1 $PASSWORD --server $YOUR_SERVER_ADDRESS

Import images
-------------

We like to use `in-place import <https://omero-guides.readthedocs.io/en/latest/upload/docs/import-cli.html#in-place-import-using-the-cli>`_ because it avoids duplicating the data on disk, but if you are only importing a small amount of data you may find non-in-place import (described below) is more convenient.
In-place import is achieved by using `import bash script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/in_place_import_as.sh>`_. Use the `publicly available data provided <https://downloads.openmicroscopy.org/images/>`_ or use your own data or `data downloaded from IDR <https://idr.openmicroscopy.org/about/download.html>`_. Shell into the machine where you have installed your OMERO.server (necessary for in-place import) and, assuming that you have for example your data in a folder ``my-folder`` which is visible from that machine::

    $ HOST=localhost SUDOER=trainer-1 PASSWORD=$PASSWORD_FOR_trainer-1 IMPORTTYPE=normal NUMBER=15 FOLDER=/path/to/my-folder bash in_place_import_as.sh

> will in-place import the images from ``my-folder`` into a new Dataset named ``/path/to/my-folder`` for user-1 through user-15.

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/ea7d1e24beaa4617ab1ce5e540624064afb32b3d

Image data can also be populated by a `Python script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py>`_ which copies images as pixeldata (i.e. not the original images) from `IDR <http://idr.openmicroscopy.org/>`_ by default. You can adjust the `appropriate line <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py#L80>`_ to copy from other OMERO.servers. Note that the script is only creating `images of single T and Z <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py#L36>`_, and thus reducing the original images dimensions in case these are multi-z or T images::

> 

Author: @will-moore : Please compare with https://github.com/ome/omero-guides/pull/236#discussion_r1044681593 - it seems there are two different schools of thought here.

> 

Author: I have replaced all the pointers to ``master`` on the ``training-scripts`` repo with ``v0.7.3`` (the ``latest`` does not work).

> 

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/ea7d1e24beaa4617ab1ce5e540624064afb32b3d

    $ python idr_copy_plate.py trainer-1 --server $YOUR_SERVER_ADDRESS $PASSWORD $PLATE_ID

> will copy the Plate with $PLATE_ID from IDR as trainer-1 into your server, but note that this ``idr_copy_plate.py`` script only works on Plates which have a single Field (Image) per Well.

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/ea7d1e24beaa4617ab1ce5e540624064afb32b3d

Import metadata
---------------

Import metadata from a CSV via `OMERO.web <https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata-ui.html>`_ or `Command Line Interface <https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata.html>`_.

Annotate images using training-scripts:

> - `Copy Key-Value pairs <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_get_map_annotations.py>`_ from IDR or other OMERO server. The command below will copy MapAnnotations from IDR to your server between images inside Projects with IDs 1 (in IDR) and 2 (in your server). The Project in your server is owned and the action is executed by `trainer-1`.::

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/ea7d1e24beaa4617ab1ce5e540624064afb32b3d

    $ python idr_get_map_annotation.py trainer-1 $PASSWORD Project:1 Project:2 --server $YOUR_SERVER_ADDRESS

- `Add new Key-Value pairs <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/key_value_pairs.py>`_. The command below will add Key-Value pairs defined inside the script randomly to the images inside Datasets with name ``big-dataset`` for all 50 users in your server. The $PASSWORD for all the users must be the same.::

    $ python key_value_pairs.py $PASSWORD big-dataset --server $YOUR_SERVER_ADDRESS

- `Calibrate images <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/calibrate_images.py>`_. The command below will calibrate all the images within Dataset named ``western-blots`` to `0.33 micrometers per pixel <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/calibrate_images.py#L72>`_ for all 50 users in your server. The $PASSWORD for all the users must be the same.::

> 

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/ea7d1e24beaa4617ab1ce5e540624064afb32b3d

Prepare a training on OMERO
===========================

This chapter describes a setup of a training OMERO.server at 

> your institution. Further, it gives overview of the other training

Reviewer 2: ```suggestion
your institution. Further, it gives an overview of the other training
```

We recommend to use the Ansible management software and use the provided Ansible playbook to install the OMERO.server for training, as this gives you the exact blueprint of the server setup used during trainings run by the OME Team.

You can use the exact `training ansible playbook <https://github.com/ome/prod-playbooks/blob/master/omero/training-server/playbook.yml>`_ used by the OME Team for their training server. Alternatively, you might start with a `simpler playbook <https://github.com/ome/ansible-example-omero-addons/blob/master/playbook.yml>`_. Both of those playbooks will also install the OMERO.web Apps for you.

You also have the option to follow the server installation steps in the `sysadmin documentation <https://omero.readthedocs.io/en/stable/sysadmins/unix/server-installation.html>`_. In this case, you will have to install `Apps for OMERO.web <https://www.openmicroscopy.org/omero/apps/>`_ in separate, post installation steps. We recommend to install all apps listed on the webpage except OMERO.gallery and OMERO.mapr. 

> In case any other installation method was used than `training ansible playbook <https://github.com/ome/prod-playbooks/blob/master/omero/training-server/playbook.yml>`_, you have to also set up the server configuration according to the `configuration block in the playbook <https://github.com/ome/prod-playbooks/blob/c72014f7f5a181d4d4daad3b86045f1c4e41a75b/omero/training-server/playbook.yml#L473>`_.

Reviewer 2: ```suggestion
In case any other installation method was used than `training ansible playbook <https://github.com/ome/prod-playbooks/blob/master/omero/training-server/playbook.yml>`_, you have to also set up the server configuration according to the `configuration block in the playbook <https://github.com/ome/prod-playbooks/blob/master/omero/training-server/playbook.yml#L47>`_.
```

Author: This whole block was removed, so the line number is not relevant now.. The `philosophical` question though remains: link to `master` or to commits ? Maybe this is related to the fact that some repos do not have frequent releases, e.g. `prod-playbooks` ?

    $ OMEUSER=trainer NUMBER=2 BULKFILE=/path/to/idr0021-experimentA-bulk.yml bash in_place_import_as.sh

to import the images for trainer-1 and trainer-2.

The `script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/in_place_import_as.sh>`_ assumes by default `in-place imports <https://omero-guides.readthedocs.io/en/latest/upload/docs/import-cli.html#in-place-import-using-the-cli>`_. You can adjust it for "classsical", non-in-place import by deleting the ``--transfer=ln_s`` from the script lines or, if using the `bulk file <https://omero-guides.readthedocs.io/en/latest/upload/docs/import-cli.html#bulk-import-using-the-cli>`_ workflow, comment out the ``transfer`` line from the bulk file.

> Image data can also be populated by a `python script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py>`_ which copies images as pixeldata (i.e. not the original images) from `IDR <http://idr.openmicroscopy.org/>`_ by default. You can adjust the `appropriate line <https://github.com/ome/training-scripts/blob/6f6866b21be4cbf48cafe2756899a21b4764b47e/maintenance/scripts/idr_copy_plate.py#L80>`_ to copy from other OMERO.servers. Note that the script is only creating `images of single T and Z <https://github.com/ome/training-scripts/blob/6f6866b21be4cbf48cafe2756899a21b4764b47e/maintenance/scripts/idr_copy_plate.py#L36>`_, and thus reducing the original images dimensions in case these are multi-z or T images::

Reviewer 2: ```suggestion
Image data can also be populated by a `Python script <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py>`_ which copies images as pixeldata (i.e. not the original images) from `IDR <http://idr.openmicroscopy.org/>`_ by default. You can adjust the `appropriate line <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py#L80>`_ to copy from other OMERO.servers. Note that the script is only creating `images of single T and Z <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_copy_plate.py#L36>`_, and thus reducing the original images dimensions in case these are multi-z or T images::
```

Import metadata from a CSV via `OMERO.web <https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata-ui.html>`_ or `Command Line Interface <https://omero-guides.readthedocs.io/en/latest/upload/docs/metadata.html>`_.

Annotate images using training-scripts:

- `Copy Key-Value pairs <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/idr_get_map_annotations.py>`_ from IDR or other OMERO server. The command below will copy MapAnnotations from IDR to your server between images inside Projects with IDs 1 (in IDR) and 2 (in your server). The Project in your server is owned and the action is executed by `trainer-1`.::

    $ python idr_get_map_annotation.py trainer-1 $PASSWORD Project:1 Project:2 --server $YOUR_SERVER_ADDRESS

- `Add new Key-Value pairs <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/key_value_pairs.py>`_. The command below will add Key-Value pairs defined inside the script randomly to the images inside Datasets with name ``big-dataset`` for all 50 users in your server. The $PASSWORD for all the users must be the same.::

    $ python key_value_pairs.py $PASSWORD big-dataset --server $YOUR_SERVER_ADDRESS

> - `Calibrate images <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/calibrate_images.py>`_. The command below will calibrate all the images within Dataset named ``western-blots`` to `0.33 micrometers per pixel <https://github.com/ome/training-scripts/blob/6f6866b21be4cbf48cafe2756899a21b4764b47e/maintenance/scripts/calibrate_images.py#L72>`_ for all 50 users in your server. The $PASSWORD for all the users must be the same.::

Reviewer 2: ```suggestion
- `Calibrate images <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/calibrate_images.py>`_. The command below will calibrate all the images within Dataset named ``western-blots`` to `0.33 micrometers per pixel <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/calibrate_images.py#L72>`_ for all 50 users in your server. The $PASSWORD for all the users must be the same.::
```

    $ python calibrate_images.py $PASSWORD western-blots --server $YOUR_SERVER_ADDRESS

> - `Add timestamps <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/set_timestamps.py>`_. The command below will set timestamps on the timelapse images within Dataset named `timestamps` with `delta T of 300 seconds <https://github.com/ome/training-scripts/blob/6f6866b21be4cbf48cafe2756899a21b4764b47e/maintenance/scripts/set_timestamps.py#L71>`_ for all 50 users in your server. The $PASSWORD for all the users must be the same.::

Reviewer 2: ```suggestion
- `Add timestamps <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/set_timestamps.py>`_. The command below will set timestamps on the timelapse images within Dataset named `timestamps` with `delta T of 300 seconds <https://github.com/ome/training-scripts/blob/master/maintenance/scripts/set_timestamps.py#L71>`_ for all 50 users in your server. The $PASSWORD for all the users must be the same.::
```

Prepare a training on OMERO
===========================

This chapter describes a setup of a training OMERO.server at 
your institution. Further, it gives overview of the other training
resources as well as hints for trainers about how to present OMERO in a training session.

Resources
---------

- Community contributions and setups: `Github issue <https://github.com/ome/omero-guides/issues/107>`_.

- `Useful scripts <https://github.com/ome/training-scripts>`_ for setting up the training server.

Spin up a training server
-------------------------

> Get an empty box or virtual machine with CentOS 7.

Author: fixed in https://github.com/ome/omero-guides/pull/236/commits/a67fecca75a54a8e47ff14ec9b94194c1677f893
